### PR TITLE
fix(replication): list projects before replicate

### PR DESCRIPTION
**Pull request by chlins**
Fix: https://github.com/goharbor/harbor/issues/15894

Signed-off-by: chlins <chenyuzh@vmware.com>

### CODE REVIEW

File: src/pkg/reg/adapter/harbor/base/adapter.go [hunk 1]
```diff
@@ -169,7 +179,28 @@ func (a *Adapter) PrepareForPush(resources []*model.Resource) error {
 			Metadata: metadata,
 		}
 	}
-	for _, project := range projects {
+
+	// get exist projects
+	q := buildProjectsQuery(projects)
+	queryProjects, err := a.Client.ListProjectsWithQuery(q, false)
+	if err != nil {
+		return fmt.Errorf("list projects with query %s error: %w", q, err)
```
Comment (wy65701436) [inline src/pkg/reg/adapter/harbor/base/adapter.go:187]: please use lib/errors for instead

File: src/pkg/reg/adapter/harbor/base/adapter.go [hunk 2]
```diff
@@ -139,6 +140,15 @@ func (a *Adapter) Info() (*model.RegistryInfo, error) {
 	return info, nil
 }
 
+func buildProjectsQuery(projects map[string]*Project) string {
```
Comment (wy65701436) [inline src/pkg/reg/adapter/harbor/base/adapter.go:143]: if it‘s not a util func, suggest using nested function.

File: src/pkg/reg/adapter/harbor/base/adapter.go (hunk 1 above)
Comment (chlins) [reply src/pkg/reg/adapter/harbor/base/adapter.go:187]: ok

File: src/pkg/reg/adapter/harbor/base/adapter.go (hunk 2 above)
Comment (chlins) [reply src/pkg/reg/adapter/harbor/base/adapter.go]: moved.

**Review by ywk253100 — APPROVED** [review]

**Review by wy65701436 — APPROVED** [review]
lgtm

**Pull request merged**